cf/test/test_Data.py

```diff
@@ -131,6 +131,293 @@ class DataTest(unittest.TestCase):
+                )
+            )
+
+        # Test masked arrays
+        # 1. Example case where the masks differ only (data is identical)
+        j1 = cf.Data(np.ma.array([1.0, 2.0, 3.0], mask=[1, 0, 0]), "m")
+        print("SELF-EQUALS J1 NOW <<<<<<<<<<<<<<<<<<<<<<<<<")
+        self.assertTrue(j1.equals(j1))
+        j2 = cf.Data(np.ma.array([1.0, 2.0, 3.0], mask=[0, 1, 0]), "m")
+        self.assertTrue(j2.equals(j2))
+        with self.assertLogs(level=cf.log_level().value) as catch:
+            self.assertFalse(j1.equals(j2))
+            self.assertTrue(
+                any(
+                    "Data: Different array values" in log_msg
+                    for log_msg in catch.output
+                )
+            )
+        # 2. Example case where the data differs only (masks are identical)
+        j3 = cf.Data(np.ma.array([1.0, 2.0, 100.0], mask=[1, 0, 0]), "m")
+        self.assertTrue(j3.equals(j3))
+        with self.assertLogs(level=cf.log_level().value) as catch:
+            self.assertFalse(j1.equals(j3))
+            self.assertTrue(
+                any(
+                    "Data: Different array values" in log_msg
+                    for log_msg in catch.output
+                )
+            )
+
+        # 3. Trivial case of data that is fully masked
+        j4 = cf.Data(np.ma.masked_all(shape, dtype="int"), "m")
+        self.assertTrue(j4.equals(j4))
+        with self.assertLogs(level=cf.log_level().value) as catch:
+            self.assertFalse(j4.equals(d))
+            self.assertTrue(
+                any(
+                    "Data: Different array values" in log_msg
+                    for log_msg in catch.output
+                )
+            )
+        # 4. Case where all the unmasked data is 'allclose' to other data but
+        # the data is not 'allclose' to it where it is masked, i.e. the data
+        # on its own (namely without considering the mask) is not equal to the
+        # other data on its own (e.g. note the 0-th element in below examples).
+        # This differs to case (2): there data differs *only where unmasked*.
+        # Note these should *not* be considered equal inside cf.Data, whereas
+        # np.ma.allclose and indeed our own _da_ma_allclose methods do hold
+        # these to be 'allclose': Data.equals is stricter than _da_ma_allclose.
+        j5 = cf.Data(np.ma.array([1.0, 2.0, 3.0], mask=[1, 0, 0]), "m")
+        self.assertTrue(j5.equals(j5))
+        j6 = cf.Data(np.ma.array([10.0, 2.0, 3.0], mask=[1, 0, 0]), "m")
+        self.assertTrue(j6.equals(j6))
+        with self.assertLogs(level=cf.log_level().value) as catch:
+            print("J5 EQUALS J6 NOW <<<<<<<<<<<<<<<<<<<<<<<<<")
+            self.assertFalse(j5.equals(j6))
+            self.assertTrue(
+                any(
+                    "Data: Different array values" in log_msg
+                    for log_msg in catch.output
+                )
+            )
```

Comment:
#### Outstanding issue 1

Currently, I can't get all of the masked array tests to pass (for known reasons but your advice would be appreciated). Either the 'extra special case' here (`self.assertFalse(j5.equals(j6))` which tests that key case which makes our equality with masked arrays differ in behaviour to `np.ma.allclose` for example) fails, when `masked_equal=True` is input to the `_da_ma_allclose`, or all of the self-equality checks on masked Data arrays fail, e.g. `self.assertFalse(j4.equals(j4))`, when we instead set `masked_equal=False`.

I wasn't sure what the appropriate choice for `masked_equal` would be, but I don't think it is the issue at hand anyway, because upon investigation I found that, with the print statement in the final debugging commit, the underlying issue is that the data comparison is comparing the data *with masked elements* (e.g. `[-- 2.0 3.0]` for the above example), and not the data extracted separately from the mask (e.g. `[1.0 2.0 3.0]`).

Surely we want to test on the latter, so need to call some method or attribute to extract the data from the mask for our `data_comparison` result? But how would we do that - I've tried using the `getdata` method but that still masks elements where the mask is `True` for the result...

As well as that seeming wrong to me, there is the issue of self-equality, for example with `masked_equal=False`, my print statements show:

```
SELF-EQUALS J1 NOW <<<<<<<<<<<<<<<<<<<<<<<<<
DATA COMP'ING IS [-- 2.0 3.0] [-- 2.0 3.0]
DATA COMP RESULT IS False
MASK COMP RESULT IS True
OVERALL COMP RESULT IS False
```
indicating that the data comparison does not xonsider `[-- 2.0 3.0]` allclose to itself due to the masked element.

What are your thoughts on this? (Ca take a video call if that seems easiest for discussion.)

Reply:
(discussing off-line - will report back later ...)

Reply:
Thanks David, all of your thoughts are much appreciated.

Reply:
This is now resolved, I believe, after discussion in other feedback comments.